pkg/agent/batch_remote_write_client.go

```diff
@@ -67,45 +79,48 @@ func (b *BatchWriteClient) Run(ctx context.Context) error {
 		case <-ctx.Done():
 			return ctx.Err()
 		case <-ticker.C:
+		case <-b.triggerBatch:
 		}
 
 		b.report(time.Now(), b.batch(ctx))
+
+		select {
+		case b.batchDone <- struct{}{}:
+		default:
+		}
 	}
 }
 
 func (b *BatchWriteClient) batch(ctx context.Context) error {
```

Comment:
```suggestion
func batch(ctx context.Contextm, r *profilestorepb.WriteRawRequest) error {
```

This method can receive the request and by this way we can get rid of the lock in this method if I'm not mistaken. This method could also omit the write client as a receiver if it's the case.